cmake: yaml: fix handling of special chars in strings

pillo79 opened the pull request:
The current YAML export has problems dealing with strings that include special characters such as single/double quotes or backslashes. This is evident when enabling the LLEXT EDK on certain targets / applications, as described in #87765. 

This PR addresses these issues by enclosing generated YAML strings in single quotes to simplify escaping. The intermediate genex file is also switched to YAML to achieve the same result.

The `to_yaml` internal function and `yaml-filter.cmake` have been improved to output the correct "flavor" of YAML depending on the current export stage.  

Finally, a new test case is added to verify that `'`, `"` and `\` are now properly processed, both when passed directly to `yaml_set()` or after genex expansion.

### Known limitations

These special characters are still not supported in the `LIST MAP` implementation. 

### Review notes
 
The first two commits in the PR series introduce no functional changes but do several renames / refactors. The rationale behind the other changes is detailed in each commit description.

pillo79 committed:
cmake: yaml: (cosmetic) avoid referring to file type in variable names

To move away from the JSON/YAML terminology, rename the variables that
refer to the intermediate files to "expanded" and "output" files.

Also, clean up a few "C-style" function calls and a typo in a variable
name in the same files.

No functional change introduced in this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
cmake: yaml: properly store strings

Store strings in YAML as single-quoted entries to avoid issues with
special characters. This also fixes a quirk with the current test
suite where the quotes in the expected value are filtered out by the
YAML import.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
cmake: yaml: switch to yaml for intermediate file

The intermediate file used to expand the generator expressions is now in
YAML format. This allows for a more robust handling of the data, as the
single quoted strings are a lot easier to escape: inside them every
character is a literal except for two single quotes (which intuitively
map to a literal single quote).

When saving a simple YAML file without genexes, the single quotes in
strings need to be escaped. However, doing so when saving the
intermediate file would make it harder to properly escape them later,
since it could be possible that the expansion of the generator
expressions would introduce new single quotes. For this reason, when
genexes are enabled, the escaping is now done in a single pass inside
the yaml-filter.cmake script.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
cmake: yaml: improve escaping of strings with backslashes and quotes

Switching the intermediate file format from JSON to YAML has a very
significant benefit: the file is now loaded via yaml_load(), which
internally calls Python to parse the file into the JSON format that
CMake expects. This means that the file contents are now automatically
escaped properly for JSON; it is a huge improvement over the previous
implementation, which was escaping everything given as input to
to_yaml().

With the removal of the now-redundant call in to_yaml(), escaping is
applied exactly once per value or list, when it is passed to yaml_set().
This allows to convert the logic in zephyr_string(ESCAPE ...) to a more
robust "escape everything" approach.

These changes fix the handling of strings with backslashes and different
types of quotes passed either directly or via generator expression. The
existing tests are updated to cover these cases.

Two other small changes are made in this commit:

- a small check in internal_yaml_list_append() is removed, as the same
  issue is already detected by the caller yaml_set() logic.

- the to_yaml() function is modified to initialize the YAML output
  variable at the top level, which is the expected behavior. This
  resulted in genex temp files sometimes having duplicate lines.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -3095,10 +3095,10 @@ function(zephyr_string)
+    # Escape every instance of '\' or '"' in the string. "\\1" is the matched
+    # character. "\\\\" is seen by CMake as \\, meaning an escaped '\', which
+    # then becomes a single '\' in the final string.
+    string(REGEX REPLACE "([\"\\])" "\\\\\\1" work_string "${ZEPHYR_STRING_UNPARSED_ARGUMENTS}")
```
would be nice if we can simplify.

iirc one reason for avoiding escaping already escaped sequences was that we need to post-process the generator expanded strings, and afaict that will still be needed.

But keep up the direction :+1:

tejlmand replied:
> > we need to post-process the generator expanded strings, and afaict that will still be needed.
>
> This should no longer be perfomed. The trick is using ' for intermediate YAML strings, so the only problematic char after expansion is the single quote

True, but the `'` char must still be properly handled, also for generator expression which are expanded at generator time, and thus handle at build time. That's what I was referring to.

But I see the `'` is not properly handled, also when using genex :+1:

tejlmand reviewed (changes requested):
Nice, but seems the single quote handling in genex is not working as expected.

If I add this diff:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3f54336dd3e..2ec248c2f2e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2301,5 +2301,6 @@ toolchain_linker_finalize()
 # export build information
 build_info(zephyr version VALUE ${PROJECT_VERSION_STR})
 build_info(zephyr zephyr-base VALUE ${ZEPHYR_BASE})
+build_info(vendor-specific my quote VALUE $<TARGET_PROPERTY:app,single_quote>)
 
 yaml_save(NAME build_info)
diff --git a/samples/hello_world/CMakeLists.txt b/samples/hello_world/CMakeLists.txt
index ecb7d24bb8f..ae5170a8db0 100644
--- a/samples/hello_world/CMakeLists.txt
+++ b/samples/hello_world/CMakeLists.txt
@@ -6,3 +6,4 @@ find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 project(hello_world)
 
 target_sources(app PRIVATE src/main.c)
+set_property(TARGET app PROPERTY single_quote "Someone's string")
```

Then it fails.
I creates a final build_info.yaml containing:
```
  vendor-specific:                                                                                  
    my:                                                                                             
      quote: 'Someone's string'              
```
which is invalid yaml.

pillo79 committed:
cmake: extensions: (cosmetic) refactor zephyr_string()

This commit refactors the zephyr_string() function to explicitly
show that ESCAPE and SANITIZE are mutually exclusive options and
that SANITIZE joins multiple arguments into a single string before
processing it.

No functional change introduced by this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 reviewed (commented):
**v2** ([comparison diff](https://github.com/zephyrproject-rtos/zephyr/compare/2915cefcbb6..5f6e5f5b429) without rebase):
- fixed bug with single quotes in genexes
- added genex escape test case
- minor refactor of `zephyr_string()` to make code paths more obvious
- simpler regex implementation for `ESCAPE`
- updated `zephyr_string` comment block
- rebased to latest CI requirements

First two commits are now cosmetic - it's easier to review them separately.

tejlmand reviewed (approved):
Thanks for this improvement :+1:
